lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs

```diff
@@ -346,22 +382,20 @@ decodeShelleyAddress
     => ByteString
     -> Either TextDecodingError Address
 decodeShelleyAddress bytes = do
+    let firstByte = BS.unpack $ BS.take 1 bytes
+
+    let knownBytes = pure <$>
+            knownDiscriminants @'Testnet ++ knownDiscriminants @'Mainnet
+    when (firstByte `notElem` knownBytes) $ Left (invalidAddressType firstByte)
+
+    let allowedBytes = pure <$> knownDiscriminants @n
+    when (firstByte `notElem` allowedBytes) $ Left invalidNetwork
+
     case BS.length bytes of
-        n | n == addrSingleSize -> do
-            let firstByte = BS.take 1 bytes
-            when (firstByte /= BS.pack [single @n]) $
-                if firstByte `elem`
-                    (BS.pack <$> [[single @'Mainnet], [single @'Testnet]])
-                    then Left invalidNetwork
-                    else Left invalidAddressType
-        n | n == addrGroupedSize -> do
-            let firstByte = BS.take 1 bytes
-            when (firstByte /= BS.pack [grouped @n]) $
-                if firstByte `elem`
-                    (BS.pack <$> [[grouped @'Mainnet], [grouped @'Testnet]])
-                    then Left invalidNetwork
-                    else Left invalidAddressType
+        n | n == addrSingleSize  && firstByte /= [addrGrouped @n] -> pure ()
+        n | n == addrGroupedSize && firstByte == [addrGrouped @n] -> pure ()
```

Comment:
I've _slightly_ re-organized this and shifted some of the logic. Before, the "length" errors would be thrown first, now the invalid address type or network are thrown first.